use rich editor for chat prompt editor

sqs opened the pull request:
Uses Facebook's [Lexical](https://lexical.dev) contenteditable editor for the chat prompt editor instead of a textarea, for:

- Higher-quality support for tokenized context mentions
- Reuse of existing typeahead code instead of implementing our own (which has some bugs with up/down arrow keys and has code to support contextQuery, selectedContext, etc., scattered across many files)
- Extensibility for more kinds of context to be mentioned inline

All existing e2e tests pass (with a few minor changes), and some new ones were added. The protocol between the webview and the extension host is unchanged.

<img width="765" alt="image" src="https://github.com/sourcegraph/cody/assets/1976/c14634fd-82b4-4797-becd-aa60cd03cbca">

Changes:

- <kbd>UpArrow</kbd> no longer populates the chat input with the contents of the previous message from history. It now edits the most recently sent message in the transcript. This new behavior is better because the previous behavior would result in a conversation that had the same question asked twice, which could lead to erroneous results. This new behavior matches Slack and other messaging apps and simplifies the implementation.

Known issues:

- If you open an old chat that was started before the rich editor was added and edit a message in the transcript that has `@`-mentions, the `@`-mentions will show up as plain text and will not actually include the mentioned files unless you re-type them (or place your cursor at the end and choose the file from the dropdown menu).


## Test plan

e2e tests

philipp-spiess reviewed (commented):
This is awesome! Big fan of Lexical and it's the perfect use case for our input.

I did find some bugs with the current implementation, though, that we might want to look into before merging.

##  UI jumping around when you insert text

I'll leave this to @toolmantim but I think constraining the size of the dropdown would help here

https://github.com/sourcegraph/cody/assets/458591/230e3ac8-eb56-4052-a621-e0a4a71930cd

## Disappearing text

When I suffix my input with ` :)` some characters go missing?

https://github.com/sourcegraph/cody/assets/458591/3a15565c-2c2d-456f-806c-7094ca0cce66

## Enter should not submit when inside the dropdown

In Slack etc., using enter when the dropdown is open will select the item (like tab). 

https://github.com/sourcegraph/cody/assets/458591/944c3124-4187-4fa9-b674-be840276c1e8

## Editing is bugged

When you edit, sometimes the @-mentions are removed. In one instance I also saw the raw vscode open URI being leaked although I was not able to repro this.

https://github.com/sourcegraph/cody/assets/458591/75380ad3-4c40-4a43-a8a2-11ed7a369f79

## Adding one inline code snippet will render a grey box next to the input

<img width="701" alt="grey box" src="https://github.com/sourcegraph/cody/assets/458591/bc4d34ee-00f2-45d3-a3a4-eb79d0ecb948">

## Symbol mentions don't work

https://github.com/sourcegraph/cody/assets/458591/b757fd85-4eda-40a9-9db5-ce3be39982c8

sqs commented on vscode/webviews/promptEditor/fixtures.ts:
```diff
@@ -0,0 +1,92 @@
+import type { SerializedLexicalNode } from 'lexical'
+import type { PromptEditorValue } from './PromptEditor'
+
+export const FILE_MENTION_VALUE_FIXTURE: PromptEditorValue = {
```
No, it is used (imported by other files).

philipp-spiess replied:
🤦 Yeah I don't know how I didn't see it, maybe I had a search filter on by mistake.

sqs commented:
@philipp-spiess Thanks for the review and for catching those bugs. Will work on those.

sqs commented:
Addressing issues from https://github.com/sourcegraph/cody/pull/3287#pullrequestreview-1913857455:

- [ ] UI jumping around when you insert text
- [ ] Disappearing text
- [ ] Enter should not submit when inside the dropdown
- [ ] Editing is bugged
- [ ] Adding one inline code snippet will render a grey box next to the input
- [ ] Symbol mentions don't work

sqs commented:
@philipp-spiess For the "Symbol mentions don't work":

They do work, but I see why it seems like they don't. (In the context files you expanded in your video, the right range is there. And see an example screenshot below.)

Previously, we converted symbol mentions to a string like `@path/to/file:12-34` in the text, and then we linked that. Now it keeps them as the text `@#SymbolName`, but it does not link them.

The real solution is to use Lexical to render messages as well as compose/edit them (and save Lexical's `editorState` in the `Transcript`. Then we will get full styling that matches what the user saw when they were composing the message. Until then, we have 2 choices:

1. Show the raw text as is (or with some hacky regexp-based styling), like `@#SymbolName`. Simpler and looks nicer in the editor, but can *look* broken in the transcript.
2. Show it like `@path/to/file:12-34` and parse those at-mentions in the `Transcript` message text, which diverges from the actual editor representation and leads to things like the bug where ` :)` at the end of a message is dropped (and other `@whatever`s in the message text being erroneously picked up).

I am trying to move away from (2). I personally would prefer (1) until we can actually save Lexical's `editorState` in the `Transcript` (hopefully soon), but if you or @toolmantim feel that would be too big of a regression, I'm happy to defer to you.

![image](https://github.com/sourcegraph/cody/assets/1976/04f3cd27-3804-46c6-a7c7-b598cd1e3df9)

sqs commented:
@philipp-spiess I am going to make `Transcript` store the Lexical `editorState` to really fix the "Editing is bugged" issue.

valerybugakov reviewed (commented):
This is beautiful!

The only concerning thing I noticed testing it locally is that when I attempt to edit the previous message multiple times (via CMD+L or arrow up) while waiting for the LLM response, the UI freezes. One time, it caused my VS Code instance to crash. It looks like we're not canceling previous requests in time, and this causes Cody to do much more work than expected.

![CleanShot 2024-03-05 at 12 44 25](https://github.com/sourcegraph/cody/assets/3846380/92a5cdea-5936-4983-bfd5-3d31c43d563a)

A couple of minor things that might be unrelated to this migration

- Scrollbar appearing and disappearing causes UI to jump:
    ![CleanShot 2024-03-05 at 12 29 04](https://github.com/sourcegraph/cody/assets/3846380/ad9097d3-db8d-41b8-bb11-06c65140a3b2)

- The search context popup is hard to read sometimes
    <img width="456" alt="Screenshot 2024-03-05 at 12 31 52" src="https://github.com/sourcegraph/cody/assets/3846380/275e8e41-f649-4964-ac67-0b6436fb1b5b">

sqs commented:
@valerybugakov Thanks!

- I can repro the freezing with `Ctrl+K then Esc` (repeated) but not with `UpArrow then Esc` (repeated). Will fix. 
- Will fix the scrollbar issue
- Will fix the `File too large` layout issue

valerybugakov commented:
> By doing this, it causes the <Chat /> component to render 10000s of times, thereby freezing up

WAT 🫣

sqs commented:
PR at https://github.com/sourcegraph/cody/pull/3314 to clean up some stuff is necessary for this PR

abeatrix commented:
Thank you for adding this! This is going to make adding additional @-supports so much easier in the future 🎉 

Here are some small additional bugs I found:

1. Does not support file range input 

Before:

![image](https://github.com/sourcegraph/cody/assets/68532117/ea68adeb-587b-45df-a787-66b6f32092d4)

![image](https://github.com/sourcegraph/cody/assets/68532117/3e17c0d5-f48e-4fa1-a154-7daed60d18d7)

After:

![image](https://github.com/sourcegraph/cody/assets/68532117/3b5137f6-2098-435c-a748-1d346f30e72f)

![image](https://github.com/sourcegraph/cody/assets/68532117/d00ec222-3807-493a-9977-f23d1ffcbd9e)

2. Hovering over the file name in Context list does not display the source of the file for @-mentioned file

Before:

![image](https://github.com/sourcegraph/cody/assets/68532117/523f5912-4c23-4c26-9902-55ecd2602efa)

After:

![image](https://github.com/sourcegraph/cody/assets/68532117/b6010c5e-0824-43d5-92c2-0afabe1f9ec8)

3. We should probably update the keybindings on `Edit Last Message` to up-arrow?

![image](https://github.com/sourcegraph/cody/assets/68532117/0f65752b-4acd-4f6c-b100-bd1f2ef7ffd1)

This is awesome!

sqs commented:
Thanks, @abeatrix!

- [ ] Does not support file range input
- [ ] Hovering over the file name in Context list does not display the source of the file for @-mentioned file
- [x] We should probably update the keybindings on Edit Last Message to up-arrow?

abeatrix commented:
> @abeatrix (cc @toolmantim):
> 
> > Does not support file range input
> 
> Do you think that we need to retain support for `@`-mentions of file _ranges_ like `@my/file.go:12-15`? I just don't imagine any users actually typing in an `@`-mention including file range line numbers. Of course mentioning symbols is useful. I can see it being useful to go from a selection in your editor in an editor to a file range (perhaps in the future `@selection` does that, which seems better than manually having the user type a file and range).

I think it'd be helpful to create code snippets from a large file or files that do not contain symbols (json, yaml, md etc), but i doubt it'd be a common need. Another thing I can think of is that it allows us to display the line numbers in the prompt for the Explain and Smell commands which users can interact with. @toolmantim can probably expand on this :)

sqs commented:
> > @abeatrix (cc @toolmantim):
> I think it'd be helpful for creating code snippets from a large file or files that do not contain symbols (json, yaml, md etc), but i doubt it'd be a common need. Another thing I can think of is that it allows us to display the line numbers in the prompt for the Explain and Smell commands which users can interact with. @toolmantim can probably expand on this :)

Cool. That is a good reason to keep it, and I realize it was just added, so it would be unfriendly to users if I removed it in this PR. Thanks! Will make sure this PR preserves this functionality.

sqs commented:
> > > @abeatrix (cc @toolmantim):
> > > I think it'd be helpful for creating code snippets from a large file or files that do not contain symbols (json, yaml, md etc), but i doubt it'd be a common need. Another thing I can think of is that it allows us to display the line numbers in the prompt for the Explain and Smell commands which users can interact with. @toolmantim can probably expand on this :)
> 
> Cool. That is a good reason to keep it, and I realize it was just added, so it would be unfriendly to users if I removed it in this PR. Thanks! Will make sure this PR preserves this functionality.

Actually @abeatrix and @toolmantim I just realized a UX problem. Usually people will hit <kbd>Tab</kbd> or <kbd>Enter</kbd> to complete an `@`-mention of a filename. It is nice if the mention is then treated as a token visually and for keyboard operations:

![image](https://github.com/sourcegraph/cody/assets/1976/b1c5cc5f-a148-4ba1-8efe-1e419d0dd2d7)

By this I mean that if you delete or use the arrow keys, the whole thing is treated as a "chip" and not just standalone characters. However, that makes it hard to use `:line-line` syntax because as soon as the user hits <kbd>Tab</kbd> or <kbd>Enter</kbd>, the token is completed and can't be "re-opened" to edit.

Possibilities:

1. Don't tokenize at-mentions. This also makes it harder to add additional UX associated with the token, such as hover information, and means that it can't contain additional hidden metadata that is not represented in its raw text (such as disambiguation infomation if there are 2 symbols with the same names, or even omitting the full path of a symbol for presentation purposes).
2. Allow them to be "re-opened". This is possible, but the UX will be weird and it will be hard to discover. For example, now after completing a token, it types a space for you, which is convenient, but that means that users would have to delete the space and then press `:` to discover the ability to use a range.
3. Let users hover over a token to see an additional action to constrain it to a range. This could work, but kinda clunky and hard to discover.
4. Don't support manual entry of file ranges. This regresses the functionality. (But as noted in the screenshot above, it does support programmatic entry of file ranges as tokens, so commands could insert them.)
5. Any other ideas?

toolmantim commented:
Great rundown of the current state.

I reckon 4 is the right call.

3 would be beneficial when users think "whoops, I need to change that" and not having to redo it, and also for discoverability. If you have a choose lines action tho, it makes sense to have a choose symbol action too.

But to take a step back, I think the primary interaction for adding files, symbols and lines should be from the file explorer tree, right click editor menus, keyboard shortcuts from the editor, or just pre-populated entries in the popover menu when you type `@`. The files are almost always already open, so needing to retype them shouldn't be needed most of the time.

sqs commented:
Cool. Option 4 is done (mostly)! 
![image](https://github.com/sourcegraph/cody/assets/1976/5f5207cd-10e5-4338-a5b3-bd63dca641c3)

sqs commented:
- [ ] Does not focus chat input when <kbd>Meta+/</kbd> is pressed

sqs commented:
OK, I've fixed all known issues and written/updated e2e tests for all the major functionality, but I still need to do some more manual testing before it's ready for review.

dominiccooney commented on vscode/test/e2e/chat-atFile.test.ts:
```diff
@@ -129,54 +116,56 @@ test.extend<ExpectedEvents>({
 
     // Ensure explicitly @-included context shows up as enhanced context
-    await expect(chatPanelFrame.getByText(/^✨ Context:/)).toHaveCount(3)
+    await expect(chatPanelFrame.getByText(/^✨ Context:/)).toHaveCount(2)
```
🤔  Two seems right. On the other hand, why did this change?

sqs replied:
Previously, <kbd>UpArrow + Enter</kbd> would send a message with the same text as the previous message you sent. Now, it edits the previous message you sent (instead of sending a new message). So, there is one less message now than before. This is in the PR description and (oops) I'll add it to the changelog as well.

dominiccooney commented on vscode/test/e2e/chat-atFile.test.ts:
```diff
@@ -188,69 +177,179 @@ test.extend<ExpectedEvents>({
+    await chatPanelFrame.getByRole('button', { name: '@buzz.ts:1-15' }).click()
+    const previewTab = page.getByRole('tab', { name: /buzz.ts, preview, Editor Group/ })
+    await previewTab.hover()
+    await expect(previewTab).toBeVisible()
```
Great tests. On reflection some helpers like fill-and-pick could cut down the couplets of typing and getting an option. But fine as is.

sqs replied:
Yep, agreed.

sqs replied:
That sounds like the kind of PR I would like to send up soon :)

dominiccooney reviewed (approved):
Looks great, what an epic.

Some minor comments inline at your discretion.

I'm curious why the context counts changed so much in some of these tests.

dominiccooney commented:
> I could implement migrations for this, but it seems like a rare case, and it lets us remove a lot of code that has seen a fair bit of bugs.

My $0.02, Lexical is quite awesome, so taking the regression (which self-heals over time anyway) is worth the unlocks for richer prompt editing. @toolmantim, WDYT?

sqs commented on vscode/webviews/promptEditor/plugins/atMentions/atMentions.tsx:
```diff
@@ -0,0 +1,200 @@
+                item.range
+                    ? `${item.range.start.line}:${item.range.start.character}-${item.range.end.line}:${item.range.end.character}`
+                    : '',
+            ].join(':')
```
Heh, same. I made it use `JSON.stringify` instead.

sqs commented on vscode/webviews/promptEditor/plugins/atMentions/OptionsList.tsx:
```diff
@@ -0,0 +1,109 @@
+    // biome-ignore lint/correctness/useExhaustiveDependencies: Intent is to run whenever `options` changes.
+    useEffect(() => {
+        // Scroll to top when options change because the prior `selectedIndex` is invalidated.
+        ref?.current?.scrollTo(0, 0)
```
It jumps, which is desired, because this is when the items have changed. We have `scroll-behavior: auto` applied to the list, which makes `scrollTo` use non-smooth scrolling.

sqs commented on vscode/webviews/promptEditor/PromptEditor.tsx:
```diff
@@ -0,0 +1,276 @@
+    /**
+     * The HTML serialization of the editor state.
+     */
+    html: string
```
There is a text serialization on the `SerializedPromptEditorValue` type, which has 3 fields: `text`, `contextItems`, `editorState`. The first 2 fields are backward-compatible (unchanged by this PR).

sqs commented on vscode/src/chat/protocol.ts:
```diff
@@ -149,11 +149,17 @@ export type ChatSubmitType = 'user' | 'user-newchat'
     submitType: ChatSubmitType
+
+    /** @see {ChatMessage.editorState} */
+    editorState?: unknown
```
Because the value is opaque to everything except the rich editor running in the webview (which this code is far away from).

Added more explanation:

```
/** An opaque value representing the text editor's state. @see {ChatMessage.editorState} */
```

ChatMessage.editorState has more info:

```
    /**
     * For messages composed in a rich text editor field, this is the representation of the editor
     * state that can be used to instantiate the editor to edit the message or to render the
     * message. This field's value is opaque to all but the rich editor, and it must validate and
     * version the value so that it can (1) support backward- and forward-compatibility and (2) fall
     * back to editing the text for invalid values.
     */
```

toolmantim commented:
Agree with @dominiccooney — I think it's low impact and easily recoverable, let's leave it as a known issue.

sqs commented on vscode/webviews/promptEditor/plugins/keyboardEvent.tsx:
```diff
@@ -0,0 +1,67 @@
+                editor.registerCommand(
+                    KEY_DOWN_COMMAND,
+                    event => {
+                        setTimeout(() =>
```
It's not necessary for escape, but it is for `KEY_DOWN_COMMAND`. I added an explanation. I don't really know /why/ it's happening, which never makes me feel good, but there is good e2e test coverage for this, so I am comfortable enough (given that this is one of a very small number of instances of uncomprehended behavior from Lexical for me).

```
                        // HACK(sqs): Without the `setTimeout` wrap, pressing UpArrow in an empty
                        // editor does not populate the editor with the contents of the last human
                        // message, and the following error is thrown: `BaseEditor.tsx:58 TypeError:
                        // Cannot assign to read only property 'dirty' of object
                        // '#<_RangeSelection>'`.
                        setTimeout(() =>
                            onKeyDown?.(event, editorSelectionStart(editor.getEditorState()) ?? 0)
                        )
```

sqs commented:
@dominiccooney:

Thank you so much for the review!

> I'm curious why the context counts changed so much in some of these tests.

The context line counts (`✨ Context: N lines from M files`) changed because of this diff, which I'll actually extract to a separate PR now:

```
--- a/vscode/webviews/chat/components/EnhancedContext.tsx
+++ b/vscode/webviews/chat/components/EnhancedContext.tsx
@@ -41,7 +41,14 @@ export const EnhancedContext: React.FunctionComponent<{
     // It checks if file.range exists first before accessing start and end.
     // If range doesn't exist, it adds 0 lines for that file.
     const lineCount = contextFiles.reduce(
-        (total, file) => total + (file.range ? file.range?.end?.line - file.range?.start?.line + 1 : 0),
+        (total, file) =>
+            total +
+            (file.range
+                ? // Don't count a line with no characters included (character == 0).
+                  (file.range.end.character === 0 ? file.range.end.line - 1 : file.range.end.line) -
+                  file.range.start?.line +
+                  1
+                : 0),
         0
```

sqs commented on vscode/webviews/promptEditor/plugins/onFocus.tsx:
```diff
@@ -0,0 +1,14 @@
+export const OnFocusPlugin: FunctionComponent<{ onFocus?: () => void }> = ({ onFocus }) => {
+    const [editor] = useLexicalComposerContext()
+
+    useLayoutEffect(() => {
```
Actually, this was just dumb code from me. It was ineffective anyway. I deleted this and moved the `window.addEventListener('focus', ...`) code to `Chat.tsx`, which has knowledge of the `isEnhancedContextOpen` state.

sqs commented on vscode/webviews/promptEditor/plugins/atMentions/atMentions.tsx:
```diff
@@ -0,0 +1,200 @@
+    if (match !== null) {
+        // The strategy ignores leading whitespace but we need to know it's
+        // length to add it to the leadOffset
+        const maybeLeadingWhitespace = match[1]
```
I copy-pasted this code from Lexical's codebase. I don't want you to think of me as someone who uses the wrong `it's` vs. `its`. :smile: 

Fixed! Thanks for the tip.

toolmantim reviewed (commented):
This is great technology! I'm excited to get more @ types in there.

I'll get some style updates in here, as I'd like to bring it back to using the styles we had currently.

There's a few regressions I spotted, which were details we had in the old implementation. Some might have been unintentional, but if they weren't let me know.

* If you hit space, it should accept the selected item (similar to the return key, and how Slack does it) to create a much more natural typing experience esp. for exact matches:

https://github.com/sourcegraph/cody/assets/153/54707ec7-7b3e-4b0a-ab37-62c74de8dd3d

* I like the new up arrow for Edit last message, but can we make the keyboard shortcut only be visible when the key will definitely be handled? e.g. For the up arrow we should only show it when the input is focused. You can see how they disappear in the current version when you lose focus from the webview:

https://github.com/sourcegraph/cody/assets/153/4b812051-429f-49fa-929c-c685ea0541be

(While we're there — it would be nice if Return to Previous Chat had an ESC shortcut, and the ESC shortcuts were only visible on input focus)

* Any changes in the token tends to make it disappear, instead of letting you edit it and reassociate, similar to our old implementation (and Slacks). This just adds a bit more friction when editing, and makes it feel like you're battling the token system a little bit more if you made a small mistake. (Slack has an interesting "broken" state for theirs, with a dotted border and question mark)

https://github.com/sourcegraph/cody/assets/153/ca7f0822-4e97-4cce-ac87-165bc2643a71

* Did you mean to remove the hint about opening files to get symbols to appear? It only mentions checking if language extensions are installed, but on big code bases they can take quite a long time to populate the VS Code symbol cache, and opening the file used to make it appear instantly. It looks like the hint is only shown for 1 char that returns zero matches, vs 2. I'd love to know if you have any ideas of how we know the loading state, and handle this all better… it can show as empty results for quite some time on sourcegraph/sourcegraph for example.

* This wasn't in the previous version, but what do you think about following VS Code’s pattern of showing the filename first, and then the path, more similar to my original design: https://www.figma.com/file/jEsnzgsf0hNuJbqB9pKaGm/VS-Code---Cody---Commands-up-front-and-ChatGPT%2B%2B?type=design&node-id=331-63779&mode=design&t=gugSSMn6GczJFc0O-4

sqs commented:
@toolmantim Thanks! On style updates, I'm force-pushing this branch to keep it rebased, so just send me the diff of styles or tell me what you had in mind, instead of you pushing to this branch.

toolmantim commented:
Accidentally entered an invalid line range, and got some interesting results 😆

![CleanShot 2024-03-14 at 16 43 01@2x](https://github.com/sourcegraph/cody/assets/153/8bce892f-38f3-4309-8dda-57e28033681c)

toolmantim commented:
> On style updates, I'm force-pushing this branch to keep it rebased, so just send me the diff of styles or tell me what you had in mind, instead of you pushing to this branch.

The hope was to match the styles before, or at least the Figma. But the actual styles were probably better as they were battle hardened against themes and small details.

sqs commented:
@toolmantim: Thanks!

- [x] If you hit space, it should accept the selected item
- [ ] can we make the UpArrow keyboard shortcut only be visible when the key will definitely be handled? e.g. For the up arrow we should only show it when the input is focused
- [ ] While we're there — it would be nice if Return to Previous Chat had an ESC shortcut, and the ESC shortcuts were only visible on input focus
- [ ] Any changes in the token tends to make it disappear, instead of letting you edit it and reassociate
- [ ] Did you mean to remove the hint about opening files to get symbols to appear?
- [ ] what do you think about following VS Code’s pattern of showing the filename first, and then the path

toolmantim commented:
> I see the value in this, but this is not going to work as tokens start to represent internal meanings instead of just the text they were typed in as. Slack has a simple case (although even there it breaks on spaces in names lol). We are already stretching this with symbol names in the text, and with docs we will get beyond what we can represent purely with plain text.

Could we just convert the token back into plain text if you attempt to edit, similar to if you copy+paste back into the field? Otherwise you're forced to do this copy+paste dance if you fat fingered the wrong option in the @ menu (or maybe I'm lazier than your average user, and just don't like retyping things from scratch… especially long path strings 😆)

https://github.com/sourcegraph/cody/assets/153/e8eeadb5-8dc2-44f0-8c91-bb75e99c4653

Or does that not for some things that don't have a plain text representation?

sqs commented:
> > I see the value in this, but this is not going to work as tokens start to represent internal meanings instead of just the text they were typed in as. Slack has a simple case (although even there it breaks on spaces in names lol). We are already stretching this with symbol names in the text, and with docs we will get beyond what we can represent purely with plain text.
> 
> Could we just convert the token back into plain text if you attempt to edit, similar to if you copy+paste back into the field? Otherwise you're forced to do this copy+paste dance if you fat fingered the wrong option in the @ menu (or maybe I'm lazier than your average user, and just don't like retyping things from scratch… especially long path strings 😆)
> 
>  CleanShot.2024-03-14.at.17.42.25.mp4 
> Or does that not work for some things that don't have a plain text representation?

Yes, we could. Full disclosure, this would be a fair bit more tricky work, so I am partially just being lazy.

- I think we may to stop showing the full paths to files once the user selects the chip. Paths can be long! So then even for files, going into "chip editing mode" inline can be jarring.
- What I DO really like about Lexical editor is that undo works well. So you can just Cmd/Ctrl+Z to undo, and you'll have the text you had before you selected the option and it got chippified. Does that solve the problem enough?

Do you think this should block merging for now while you play around with it and think?

sqs commented:
@toolmantim I've updated the styling. PTAL. Let me know if I missed anything. I asked a couple questions in the Figma.

![image](https://github.com/sourcegraph/cody/assets/1976/707d637f-d776-46a8-98b5-ee741a012cc8)

toolmantim reviewed (approved):
New styles look good! I'll make a few small tweaks in a follow up PR, but there's no blockers here from me 🚀

btw undo seems a bit funky — it doesn't quite tick the box for helping people go back into token-choosing mode I think (though maybe that's just bugs/quirks getting in the road):

https://github.com/sourcegraph/cody/assets/153/433c9dbd-652f-4045-9dcd-cd32b3c97bb8

(edit: though undo after hitting enter/space does seem to work well, though chops a character off each time)

philipp-spiess reviewed (approved):
Love this so much! Thank you!
